Windows support

mthiesen opened the pull request:
@tmfink: As you suggested I have created a feature branch and created a new pull request from that. It is not possible to change the source branch after a pull request has been created.

--------

I added Windows compilation support. I used the [*gcc*](https://crates.io/crates/gcc) crate instead of *CMake* or the *Visual Studio* solution provided by *Capstone*. This has the following advantages:

- No dependency on *CMake* or *Visual Studio*. This ensures that there are no additional build requirements. If you can build regular *Rust* programs, you can build this crate.
- Both *GNU* and *MSVC* flavors of the Rust toolchain work without any changes.
- The `crt-static` *Rust* feature flag is handled correctly, when it is set the *MSVC* compiler is invoked with `/MT` instead of `/MD`.

The downside to this approach is of course that there is another build script to maintain. If the *Capstone* lib changes the Windows build may break.

I have tested `examples/demo.rs` from the `capstone-rs` crate with the following toolchains:

- `stable-x86_64-pc-windows-msvc`
- `stable-i686-pc-windows-msvc`
- `stable-x86_64-pc-windows-gnu`

mthiesen committed:
Windows support using the cc crate

tmfink commented:
Looks like tests fail (which is the problem I was running into):
~~~
$ cargo test
    Finished dev [unoptimized + debuginfo] target(s) in 0.0 secs
     Running target\debug\deps\capstone_sys-4b9bbef9612e834d.exe

running 53 tests
test bindgen_test_layout___va_list_tag ... FAILED
test bindgen_test_layout_arm64_op_mem ... ok
test bindgen_test_layout_arm_op_mem ... ok
test bindgen_test_layout_cs_arm ... ok
test bindgen_test_layout_cs_arm64 ... ok
test bindgen_test_layout_cs_arm64_op ... ok
test bindgen_test_layout_cs_arm64_op__bindgen_ty_1 ... ok
test bindgen_test_layout_cs_arm64_op__bindgen_ty_2 ... ok
test bindgen_test_layout_cs_arm_op ... ok
test bindgen_test_layout_cs_arm_op__bindgen_ty_1 ... ok
test bindgen_test_layout_cs_arm_op__bindgen_ty_2 ... ok
test bindgen_test_layout_cs_detail ... ok
test bindgen_test_layout_cs_detail__bindgen_ty_1 ... ok
test bindgen_test_layout_cs_insn ... FAILED
test bindgen_test_layout_cs_mips ... ok
test bindgen_test_layout_cs_mips_op ... ok
test bindgen_test_layout_cs_mips_op__bindgen_ty_1 ... ok
test bindgen_test_layout_cs_opt_mem ... FAILED
test bindgen_test_layout_cs_opt_skipdata ... FAILED
test bindgen_test_layout_cs_ppc ... ok
test bindgen_test_layout_cs_ppc_op ... ok
test bindgen_test_layout_cs_ppc_op__bindgen_ty_1 ... ok
test bindgen_test_layout_cs_sparc ... ok
test bindgen_test_layout_cs_sparc_op ... ok
test bindgen_test_layout_cs_sparc_op__bindgen_ty_1 ... ok
test bindgen_test_layout_cs_sysz ... ok
test bindgen_test_layout_cs_sysz_op ... ok
test bindgen_test_layout_cs_sysz_op__bindgen_ty_1 ... ok
test bindgen_test_layout_cs_x86 ... ok
test bindgen_test_layout_cs_x86_op ... ok
test bindgen_test_layout_cs_x86_op__bindgen_ty_1 ... ok
test bindgen_test_layout_cs_xcore ... ok
test bindgen_test_layout_cs_xcore_op ... ok
test bindgen_test_layout_cs_xcore_op__bindgen_ty_1 ... ok
test bindgen_test_layout_mips_op_mem ... ok
test bindgen_test_layout_ppc_op_crx ... ok
test bindgen_test_layout_ppc_op_mem ... ok
test bindgen_test_layout_sparc_op_mem ... ok
test bindgen_test_layout_sysz_op_mem ... ok
test bindgen_test_layout_x86_op_mem ... ok
test bindgen_test_layout_xcore_op_mem ... ok
test test::test_arch_arm ... ok
test test::test_arch_arm64 ... ok
test test::test_arch_mips ... ok
test test::test_arch_ppc ... ok
test test::test_arch_sparc ... ok
test test::test_arch_sysz ... ok
test test::test_arch_x86 ... ok
test test::test_arch_xcore ... ok
test test::test_cs_support ... ok
test test::test_cs_version ... ok
test test::test_non_arch_types ... ok
test test::test_x86_disassembly ... ok

failures:

---- bindgen_test_layout___va_list_tag stdout ----
	thread 'bindgen_test_layout___va_list_tag' panicked at 'assertion failed: `(left == right)`
  left: `16`,
 right: `24`: Size of: __va_list_tag', target\debug\build\capstone-sys-346b4f8e4123ac85\out/capstone.rs:9016:4
note: Run with `RUST_BACKTRACE=1` for a backtrace.

---- bindgen_test_layout_cs_insn stdout ----
	thread 'bindgen_test_layout_cs_insn' panicked at 'assertion failed: `(left == right)`
  left: `232`,
 right: `240`: Size of: cs_insn', target\debug\build\capstone-sys-346b4f8e4123ac85\out/capstone.rs:8602:4

---- bindgen_test_layout_cs_opt_mem stdout ----
	thread 'bindgen_test_layout_cs_opt_mem' panicked at 'assertion failed: `(left == right)`
  left: `20`,
 right: `40`: Size of: cs_opt_mem', target\debug\build\capstone-sys-346b4f8e4123ac85\out/capstone.rs:100:4

---- bindgen_test_layout_cs_opt_skipdata stdout ----
	thread 'bindgen_test_layout_cs_opt_skipdata' panicked at 'assertion failed: `(left == right)`
  left: `12`,
 right: `24`: Size of: cs_opt_skipdata', target\debug\build\capstone-sys-346b4f8e4123ac85\out/capstone.rs:189:4


failures:
    bindgen_test_layout___va_list_tag
    bindgen_test_layout_cs_insn
    bindgen_test_layout_cs_opt_mem
    bindgen_test_layout_cs_opt_skipdata

test result: FAILED. 49 passed; 4 failed; 0 ignored; 0 measured; 0 filtered out

error: test failed, to rerun pass '--lib'
~~~

I'm willing accept this PR if the tests fail. However, we need to document that the crate builds on Windows *but* the crate is not supported on Windows (per issue #15).

tmfink commented on build.rs:
```diff
@@ -119,44 +121,123 @@ fn main() {
         }
         header_search_paths.push(PathBuf::from("capstone/include"));
-        link_type = Some(LinkType::Static);
+        link_type = Some(LinkType::Static);        
```
You should get rid of this trailing whitespace.

mthiesen replied:
Done!

tmfink commented on build.rs:
```diff
@@ -119,44 +121,123 @@ fn main() {
+                    target_features.split(",").any(|f| f == "crt-static")
+                };
+
+                let files = [
```
Can we automate this list generation? Even adding a bash script to [`scripts/`](scripts) would be preferable. When we upgrade the underlying Capstone C library, a script will make it easier (even if we need to modify the script).

tmfink replied:
I didn't notice the script you used to generate this list.  =)

tmfink commented on Cargo.toml:
```diff
@@ -22,6 +22,9 @@ bindgen = { optional = true, version = "0.30.0" }
 pkg-config = { optional = true, version = "0.3" }
 cmake = {optional = true, version = "0.1.2"}
 
+[target.'cfg(windows)'.build-dependencies]
```
Can you put the use of the `cc` crate behind a feature? On Windows, we could enable the feature by default. Users of other OSs may also want to build Capstone with the `cc` crate.

mthiesen replied:
Unfortunately `cargo` currently does not allow default features to vary based on the platform. This means `cc` needs to be a non-option build-dependency. Do you have a better idea?

tmfink replied:
Good point. You could make a `build_capstone_cc` feature that depends on the `cc` crate.
On Windows, the `cc` crate will always be a dependency. However, on other platforms, users can still opt into the feature.

Example:
https://github.com/tmfink/capstone-sys/commit/9f80d6020e252f4f3228894d09c1e4258836954d

tmfink reviewed (changes requested):
Could you implement these changes and squash your commits?

mthiesen commented:
I am pretty sure that the failed tests are a non-issue. I ran the tests for the following toolchains:

| Toolchain | Test results |
| --------- | ------------ |
| `stable-i686-pc-windows-gnu` | 49/53 |
| `stable-i686-pc-windows-msvc` | 49/53 |
| `stable-x86_64-pc-windows-gnu` | 53/53 |
| `stable-x86_64-pc-windows-msvc` | 53/53 |

I also tried `stable-i686-apple-darwin` and `stable-i686-unknown-linux-gnu` but they didn't build for me out of the box.

As you can see the tests for the x86_64 toolchains succeed while the i686 toolchains fail. I think this is due to the fact that sizes and offsets in the generated tests are hardcoded and do not take into account varying pointer sizes. There is no way that the same tests work for 32 bit pointers and 64 bit pointers. Also note that all tests that fail are tests for `struct`s that contain pointers.

See also [this bindgen issue](https://github.com/rust-lang-nursery/rust-bindgen/issues/1213).

tmfink commented on build.rs:
```diff
@@ -119,41 +121,120 @@ fn main() {
+            }
+
+            #[cfg(not(windows))]
+            {
```
Could we also refactor this block into a separate function? Something like `build_capstone_gmake()`?

tmfink commented on build.rs:
```diff
@@ -119,41 +121,120 @@ fn main() {
-
-            println!("cargo:rustc-link-search=native={}", out_dir);
+            #[cfg(windows)]
+            {
```
Could we refactor this block into a separate function? Something like `build_capstone_cc()`?

(also see comment below for other refactor)

tmfink reviewed (changes requested):
Thanks for your patience!

tmfink commented:
I was not aware of issue rust-lang-nursery/rust-bindgen#1213. I filed issue #18 to handle the issue (which does not need to be handled by this PR). I agree that the unit tests should not need to pass (for now). 

I did not even bother to "test" my earlier attempt at Windows support because I saw the same tests failed. Looks like I may have already gotten it working...

Anyway, please handle the rest of my review comments. Thanks for the great work!

mthiesen commented:
This should be it. I have done the following:

- Moved the build code into three individual functions, one for each build tool: `build_capstone_gmake()`, `build_capstone_cmake()` and `build_capstone_cc()`.
- Generate the file list for the `cc` build on-the-fly (see: `build_capstone_cc()`).
- Introduced a new feature called `build_capstone_cc`.
- Removed some instances where subfolders of the capstone directory where hardcoded (e.g. `capstone/include`). Now everywhere `CAPSTONE_DIR` ist used.

To make sure that everything works as expected, I ran `cargo test` for all build methods on all toolchains that I have access to. This is the result:

| Toolchain                       | sys   | gmake | cmake | cc    |
| ------------------------------- | ----- | ----- | ----- | ----- |
| `stable-x86_64-apple-darwin`      | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: |
| `stable-i686-apple-darwin`        | :x:<sup>1</sup>    | :x:<sup>2</sup>    | :heavy_check_mark:* | :heavy_check_mark:* |
| `stable-x86_64-pc-windows-msvc`   | :no_entry_sign:     | :no_entry_sign:     | :no_entry_sign:     |  :heavy_check_mark: |
| `stable-x86_64-pc-windows-gnu`    | :no_entry_sign:     | :no_entry_sign:     | :no_entry_sign:     |  :heavy_check_mark: |
| `stable-i686-pc-windows-msvc`     | :no_entry_sign:     | :no_entry_sign:     | :no_entry_sign:     |  :heavy_check_mark:* |
| `stable-i686-pc-windows-gnu`      | :no_entry_sign:     | :no_entry_sign:     | :no_entry_sign:     |  :heavy_check_mark:* |
| `stable-x86_64-unknown-linux-gnu` |  :heavy_check_mark: |  :heavy_check_mark: |  :heavy_check_mark: |  :heavy_check_mark: |
| `stable-i686-unknown-linux-gnu` | :x:<sup>1</sup>    | :x:<sup>2</sup>    | :heavy_check_mark:* | :heavy_check_mark:* |

| Symbol | Meaning |
| ------- | --------- |
|  :heavy_check_mark: | build succeeds, all tests pass |
|  :heavy_check_mark:* | build succeeds, some `bindgen` tests fail |
| :no_entry_sign: | build method is not supported |
| :x:<sup>1</sup> | build fails, `brew` or `apt-get` only install `x86_64` versions of `libcapstone` |
| :x:<sup>2</sup> | build fails, `gmake` only builds `x86_64` library |

Note that none of the failures are a result of my changes. They happend before as well.

tmfink commented on Cargo.toml:
```diff
@@ -21,6 +21,9 @@ travis-ci = { repository = "capstone-rust/capstone-sys" }
 pkg-config = { optional = true, version = "0.3" }
 cmake = {optional = true, version = "0.1.2"}
+cc = {optional = true, version = "1.0"}
+[target.'cfg(windows)'.build-dependencies]
```
Could you put a newline here to separate the `[target.'cfg(windows)'.build-dependencies]` line from the previous line?

tmfink replied:
No worries, I'll make the minor change myself.

tmfink commented:
@mthiesen nice job, thanks for the good work.
